### PR TITLE
fix: require statements ignored if not 'true' requires. fixes #1142

### LINKED ISSUE #1142: Quantum not working well with variables named "require" inside AngularJS code
Greetings,
I'm having an issue very similar to this one: https://github.com/fuse-box/fuse-box/issues/1133
In AngularJS (1.6.9) there is this piece of code:

```
[...]
    function getControllers(directiveName, require, $element, elementControllers) {
      var value;

      if (isString(require)) {
        var match = require.match(REQUIRE_PREFIX_REGEXP);
        var name = require.substring(match[0].length);
[...]
```

Which doesn't get parse correctly when using quantum. 

Error: `ReferenceError: require is not defined`

Converted code (without uglify):  

```
function getControllers(directiveName, $fsx, $element, elementControllers) {
                    var value;
                    if (isString($fsx)) {
                        var match = require.match(REQUIRE_PREFIX_REGEXP);
                        var name = require.substring(match[0].length);

```

Reading the issue linked above i updated fusebox to 3.2.0-next10 but then something similar happened to another piece of code in AngularJS:

AngularJS code:

```
  function getDirectiveRequire(directive) {
    var require = directive.require || (directive.controller && directive.name);

    if (!isArray(require) && isObject(require)) {
      forEach(require, function(value, key) {
        var match = value.match(REQUIRE_PREFIX_REGEXP);
        var name = value.substring(match[0].length);
        if (!name) require[key] = match[0] + key;
      });
    }
```
Error: `TypeError: e.match is not a function`

Converted code (without uglify):  

```
function getDirectiveRequire(directive) {
            var require = directive.require || directive.controller && directive.name;
            if (!isArray($fsx) && isObject($fsx)) {
                forEach($fsx, function (value, key) {
                    var match = value.match(REQUIRE_PREFIX_REGEXP);
                    var name = value.substring(match[0].length);
                    if (!name)
                        require[key] = match[0] + key;
                });
            }
            return $fsx;
        }
```

This is my quantum configuration (i have tried with any combination of replaceTypeOf and mangle, also completely disabling uglify, still not working)
```

				this.isProduction && QuantumPlugin({
					//bakeApiIntoBundle: "app",
					replaceTypeOf: false,
					treeshake: true,
					uglify: {
						mangle: true
					},
					css: {
						clean: true
					},
					api: core => {
						core.solveComputed("moment/moment.js", {
							mapping: "moment/locale**",
							fn: statement => {
                                                             statement.setExpression(`"moment/locale/" + name + ".js"`);
							}
						});
					}
					// containedAPI : true
				})
```

### PR DESCRIPTION
**Pull request by jpike88**
fixes #1142

### CODE REVIEW

File: src/quantum/core/AstUtils.ts [hunk 1]
```diff
@@ -232,8 +232,19 @@ export function matchesSingleVariable(node: any, name: string) {
 }
 
 
-export function matchesRequireFunction(node: any) {
-    return matchesSingleVariable(node, "require");
+export function isTrueRequireFunction(node: any) {
+    if (matchesSingleVariable(node, "require")) {
```
Comment (nchanged) [inline src/quantum/core/AstUtils.ts]: I think here should be something like:
```
if( node.$parent.skipRequireSubstitution ) {
  node.skipRequireSubstitution = true
}
```

in order to keep it nested

**Comment by jpike88** [comment]
it's just been given a solid makeover, have another look

File: src/quantum/tests/GlobalVariable.test.ts [hunk 2]
```diff
@@ -74,6 +74,60 @@ export class GlobalVariableTEst {
...
+                    `
+                    },
+                    plugins: [QuantumPlugin({
+                        target: "browser"
+                    })]
+                }
+            }
+        ).simple().then(test => test.browser((window, env) => {
+            const app = env.getScript("app.js");
+            app.shouldFindString("require");
```
Comment (nchanged) [inline src/quantum/tests/GlobalVariable.test.ts]: `isString(require)` ?

File: src/quantum/tests/GlobalVariable.test.ts [hunk 3]
```diff
@@ -74,6 +74,60 @@ export class GlobalVariableTEst {
...
```
Comment (nchanged) [inline src/quantum/tests/GlobalVariable.test.ts]: app.shouldFindString("function(require)");
app.shouldFindString("isString(require)");
app.shouldNotFindString("$fsx");

File: src/quantum/tests/GlobalVariable.test.ts [hunk 4]
```diff
@@ -140,4 +194,29 @@ export class GlobalVariableTEst {
...
+                    `
+                    },
+                    plugins: [QuantumPlugin({
+                        target: "browser"
+                    })]
+                }
+            }
+        ).simple().then(test => test.browser((window, env) => {
+            const app = env.getScript("app.js");
+            app.shouldNotFindString("require('test')");
```
Comment (nchanged) [inline src/quantum/tests/GlobalVariable.test.ts]: You are executing an invalid js here ;-) require is `2` should be a function instead

File: src/quantum/tests/GlobalVariable.test.ts [hunk 5]
```diff
@@ -140,4 +194,29 @@ export class GlobalVariableTEst {
...
+        }));
+    }
+
 }
```
Comment (nchanged) [inline src/quantum/tests/GlobalVariable.test.ts]: Let's add one more
```
const require = function(str){ return str };
const hello = {
	foo  : {
		bar : {
			woo : () => return require("foobar")
			
		}
	}
}
	

module.exports = hello.foo.bar.woo()

```

and assertion

```
 const res = window.$fsx.r(0);
should(res).equal("foobar")
```

Comment (jpike88) [reply src/quantum/tests/GlobalVariable.test.ts]: this test fails in a weird way, it's out of the scope of what i'm fixing...

File: src/quantum/tests/GlobalVariable.test.ts (hunk 4 above)
Comment (jpike88) [reply src/quantum/tests/GlobalVariable.test.ts]: screen shot 2018-04-27 at 9 03 02 pm

File: src/quantum/tests/GlobalVariable.test.ts [hunk 6]
```diff
@@ -74,6 +74,63 @@ export class GlobalVariableTEst {
     }
 
 
+    public "Should replace require statement 2"() {
```
Comment (nchanged) [inline src/quantum/tests/GlobalVariable.test.ts]: nitpick: should NOT

File: src/quantum/tests/GlobalVariable.test.ts [hunk 7]
```diff
@@ -74,6 +74,63 @@ export class GlobalVariableTEst {
...
+                    })]
+                }
+            }
+        ).simple().then(test => test.browser((window, env) => {
+            const app = env.getScript("app.js");
+            app.shouldFindString("isString(require)");
+        }));
+    }
+
+    public "Should replace require statement 3"() {
```
Comment (nchanged) [inline src/quantum/tests/GlobalVariable.test.ts]: Should NOT

File: src/quantum/tests/GlobalVariable.test.ts [hunk 8]
```diff
@@ -140,4 +197,60 @@ export class GlobalVariableTEst {
...
+                    `
+                    },
+                    plugins: [QuantumPlugin({
+                        target: "browser"
+                    })]
+                }
+            }
+        ).simple().then(test => test.browser((window, env) => {
+            const app = env.getScript("app.js");
+            app.shouldNotFindString("require(2)");
```
Comment (nchanged) [inline src/quantum/tests/GlobalVariable.test.ts]: I think it should replace it? Because it was overridden earlier

**Comment by nchanged** [comment]
@jpike88 Thank you very much for contributing and spending so much time polishing up this PR! ❤️

**Comment by jpike88** [comment]
A project as awesome as fuse-box deserves everyone's love. I don't usually like to compliment things but seriously you deserve big kudos for this. I'll be around for other improvement and touch ups if you need the help!

**Pull request merged**